libs/jwst-codec/src/codec/integer.rs

```diff
@@ -0,0 +1,185 @@
+use super::*;
+use byteorder::WriteBytesExt;
+use nom::{number::complete::be_u8, Needed};
+use std::io::{Error, Write};
+
+pub fn read_var_u64(input: &[u8]) -> IResult<&[u8], u64> {
+    let mut shift = 0;
+    let mut result: u64 = 0;
+    let mut rest = input;
+    loop {
+        let (tail, byte) = match be_u8(rest) {
+            Ok(t) => t,
+            Err(nom::Err::Incomplete(Needed::Size(n))) => {
+                return Err(nom::Err::Incomplete(Needed::Size(
+                    n.checked_add(1).unwrap(),
+                )))
+            }
+            Err(err) => return Err(err),
+        };
+        let byte_val = byte as u64;
+        result |= (byte_val & 0b0111_1111) << shift;
+        if byte_val & 0b1000_0000 == 0 {
+            return Ok((tail, result));
+        }
+        shift += 7;
+        rest = tail;
+    }
+}
+
+pub fn write_var_u64<W: Write>(buffer: &mut W, mut num: u64) -> Result<(), Error> {
+    // bit or 0b1000_0000 pre 7 bit if has more bits
+    while num >= 0b10000000 {
+        buffer.write_u8(num as u8 & 0b0111_1111 | 0b10000000)?;
+        num >>= 7;
+    }
+
+    buffer.write_u8((num & 0b01111111) as u8)?;
+
+    Ok(())
+}
+
+pub fn read_var_i64(i: &[u8]) -> IResult<&[u8], i64> {
+    // parse the first byte
+    let (i, byte1) = be_u8(i)?;
+    // get the sign bit and the first 6 bits of the number
+    let sign_bit = (byte1 >> 6) & 0b1;
+    let mut num = (byte1 & 0b0011_1111) as i64;
+
+    let mut shift = 6;
+    let mut curr_byte = byte1;
+    let mut rest = i;
+    loop {
+        // If the sign bit is set, we need more bits
+        if curr_byte & 0b1000_0000 != 0 {
+            // Parse the next byte
+            let (tail, next_byte) = match be_u8(rest) {
```

Comment:
create a `fn map_be_u8_err` to reuse `map_err` logic: `be_u8(rest).map_err(map_be_u8_err)?`;